queasars/utility/spsa_termination.py

```diff
@@ -0,0 +1,143 @@
+
+from time import time
+from typing import Optional
+from numpy.typing import NDArray
+import logging
+from sys import stdout
+
+
+class SPSATerminationChecker:
+    """
+    Termination checker for qiskit_algorithms SPSA optimizer. For each iteration, it checks that the
+    absolute difference between the last and new function value, divided by the last function value,
+    does not fall below minimum_relative_change for more than allowed_consecutive_violations times.
+    This termination checker also keeps track of the best function value and parameter values
+    seen during the optimization, as well as the entire history of observed function values.
+
+    :param minimum_relative_change: threshold value for the change in function value relative to the last
+        function value, below which this termination checker chooses to terminate the SPSA optimization
+    :type minimum_relative_change: float
+    :param allowed_consecutive_violations: determines how often the threshold value can be violated consecutively
+        before this termination checker chooses to terminate. If set to 0, this terminates the first time the change
+        falls below the threshold value. If set to 2 for example, this terminates the first time the change falls below
+        the threshold three consecutive times. Must be at least 0.
+    :type allowed_consecutive_violations: int
+    :param maxfev: maximum amount of function evaluations the SPSA optimizer may utilize
+    :type maxfev: Optional[int]
+    """
+
+    def __init__(
+        self,
+        minimum_relative_change: float,
+        allowed_consecutive_violations: int,
+        maxfev: Optional[int] = None,
+    ):
+        self._minimum_relative_change: float = minimum_relative_change
+        self._allowed_consecutive_violations: int = allowed_consecutive_violations
+        self._maxfev: Optional[int] = maxfev
+        self._function_value_history: list[float] = []
+        self._change_history: list[float] = []
+        self._n_function_evaluations = 0
+        self._n_function_evaluation_history: list[float] = []
+        self._best_function_value: float = float("inf")
+        self._best_parameter_values: Optional[NDArray] = None
+        self._done: bool = False
+
+    def termination_check(
+        self,
+        n_function_evaluations: int,
+        parameter_values: NDArray,
+        function_value: float,
+        step_size: float,
+        accepted: bool,
+    ) -> bool:
+        """Given the callback values provided by qiskit_algorithm's SPSA optimizer, this method determines
+        whether the SPSA optimization should terminate"""
+
+        if self._done:
+            self._function_value_history = []
+            self._change_history = []
+            self._n_function_evaluations = 0
+            self._n_function_evaluation_history = []
+            self._best_function_value = float("inf")
+            self._best_parameter_values = None
+            self._done = False
```

Comment:
What does `_done` specify? Do you use this as a reset because it's a single instance that's being somehow reused by SPSA? Or over multiple concurrent SPSA-runs?

Reply:
the `qiskit_algorithms` `SPSA` object is stateless, and could be reused. Since the callback that is now part of the SPSA algorithm cannot be stateless if it investigates stability of function evaluations, it must somehow be able to reset itself if the SPSA object is reused.

Reply:
Ok, right it's not stateless, but what is its life time? I would have assumed that it's only used for a single SPSA optimization run (until it is finished) and then thrown away, so there is no need to reset it.

But of course, if you give an instance of this termination criterion to another class that runs SPSA multiple times with it, you have to reset it. Is that the way it's being used?

Reply:
In most cases, EVQE deep copies the optimizer when using it. This should avoid that error. It currently only reuses it within the mutation to optimize all layers of an individual. But in that instance, I could also just deep copy the optimizer as well. Therefore, I can see to it, that in EVQE the lifetime of the SPSA object is one optimization, if you think that this should be the intended behaviour.

Reply:
I see, no worries then, you can keep it as is I think.